// raven.database/index.js
const functions = require("./functions.js");
const langs = ["tr", "en"];
const fs = require("fs");

class RavenDB {
  constructor(options) {

    this.dbName = options["dbName"];
    this.dbPath = options["dbPath"];
    this.noBlankData = options["noBlankData"] ? (typeof options["noBlankData"] === "boolean" ? options["noBlankData"] : false) : false;
    this.readable = options["readable"] ? (typeof options["readable"] === "boolean" ? true : false) : false;
    this.lang = options["language"] ? (langs.includes(options["language"].toLowerCase()) ? options["language"].toLowerCase() : "en") : "en";
    this.message = require(`./language/${this.lang.toLowerCase()}.json`);

    functions.fetchFiles(this.dbPath, this.dbName);
  }
  
  set(db, data) {
    functions.fetchFiles(this.dbPath, this.dbName);

    if(!db) {
      throw new TypeError(this.message["errors"]["blankName"]);
    }

    if(!data) {
      throw new TypeError(this.message["errors"]["blankData"]);
    }

    const content = JSON.parse(fs.readFileSync(`./${this.dbPath}/${this.dbName}.json`, "utf8"));
    functions.set(db, data, content);

    if(this.readable) {
      fs.writeFileSync(`./${this.dbPath}/${this.dbName}.json`, JSON.stringify(content, null, 2));
    } else {
      fs.writeFileSync(`./${this.dbPath}/${this.dbName}.json`, JSON.stringify(content));
    }
    return this.get(db);
    
  }

  get(db) {

    if(!db) {
      throw new TypeError(this.message["errors"]["blankName"]);
    }

    const content = JSON.parse(fs.readFileSync(`./${this.dbPath}/${this.dbName}.json`, "utf8"));

    return functions.get(content, ...db.split("."));

  }

  fetch(db) {

    if(!db) {
      throw new TypeError(this.message["errors"]["blankName"]);
    }

    const content = JSON.parse(fs.readFileSync(`./${this.dbPath}/${this.dbName}.json`, "utf8"));

    return functions.get(content, ...db.split("."));

  }

  has(db) {

    if(!db) {
      throw new TypeError(this.message["errors"]["blankName"]);
    }

    const content = JSON.parse(fs.readFileSync(`./${this.dbPath}/${this.dbName}.json`, "utf8"));

    return functions.get(content, ...db.split(".")) ? true : false;

  }

  delete(db) {
    functions.fetchFiles(this.dbPath, this.dbName);

    if(!db) {
      throw new TypeError(this.message["errors"]["blankName"]);
    }

    const content = JSON.parse(fs.readFileSync(`./${this.dbPath}/${this.dbName}.json`, "utf8"));

    if(!this.get(db)) {
      return false;
    }

    functions.delete(content, db);

    if(this.noBlankData === true) {
      functions.removeEmptyData(content);
    }
    
    if(this.readable) {
      fs.writeFileSync(`./${this.dbPath}/${this.dbName}.json`, JSON.stringify(content, null, 2));
    } else {
      fs.writeFileSync(`./${this.dbPath}/${this.dbName}.json`, JSON.stringify(content));
    }

    return true;
  }

  add(db, number) {
    
    if(!db) {
      throw new TypeError(this.message["errors"]["blankName"]);
    }

    if(!number) {
      throw new TypeError(this.message["errors"]["blankData"]);
    }

    if(isNaN(number)) {
      throw new TypeError(this.message["errors"]["blankNumber"]);
    }

    this.set(db, Number(this.get(db) ? (isNaN(this.get(db)) ? Number(number) : this.get(db)+Number(number)) : Number(number)));
    
    return this.get(db);

  }

  subtract(db, number) {
    
    if(!db) {
      throw new TypeError(this.message["errors"]["blankName"]);
    }

    if(!number) {
      throw new TypeError(this.message["errors"]["blankData"]);
    }

    if(isNaN(number)) {
      throw new TypeError(this.message["errors"]["blankNumber"]);
    }

    if(this.get(db)-number <= 1) {
      this.delete(db);
      return (this.get(db) || 0)
    }

    if(!this.get(db)) {
      this.delete(db);
      return (this.get(db) || 0)
    }

    this.set(db, this.get(db) ? (this.get(db)-Number(number) <= 1 ? 1 : (isNaN(this.get(db)) ? 1 : this.get(db)-Number(number)) || 1) : 1);
    
    return this.get(db);

  }

  push(db, data) {

    if(!db) {
      throw new TypeError(this.message["errors"]["blankName"]);
    }

    if(!data) {
      throw new TypeError(this.message["errors"]["blankData"]);
    }

    var arr = [];

    if(this.get(db)) {
      if(typeof this.get(db) !== "object") {
        arr = [];
      } else {
        arr = this.get(db);
      }
    }

    arr.push(data);

    this.set(db, arr);

    return this.get(db);

  }

  unpush(db, data) {

    if(!db) {
      throw new TypeError(this.message["errors"]["blankName"]);
    }

    if(!data) {
      throw new TypeError(this.message["errors"]["blankData"]);
    }

    var arr = [];

    if(this.get(db)) {
      arr = this.get(db);
    }

    arr = arr.filter((x) => x !== data);

    this.set(db, arr);

    return this.get(db);

  }

  getAll() {
    const content = JSON.parse(fs.readFileSync(`./${this.dbPath}/${this.dbName}.json`, "utf8"));

    return content;
  }

  deleteAll() {

    fs.writeFileSync(`./${this.dbPath}/${this.dbName}.json`, JSON.stringify({}, null, 2));

    return true;

  }

}

module.exports = RavenDB;
